go/vt/vttablet/tabletmanager/semisyncmonitor/monitor.go

```diff
@@ -0,0 +1,390 @@
+	// if the primary is blocked on semi-sync ACKs or not.
+	ticks *timer.Timer
+	// clearTicks is the ticker to clear the data in
+	// the semisync_recover table.
+	clearTicks *timer.Timer
+
+	// mu protects the fields below.
+	mu      sync.Mutex
+	appPool *dbconnpool.ConnectionPool
+	isOpen  bool
+	// isWriting stores if the monitor is currently writing to the DB.
+	// We don't want two different threads initiating writes, so we use this
+	// for synchronization.
+	isWriting bool
+	// inProgressWriteCount is the number of writes currently in progress.
+	// The writes from the monitor themselves might get blocked and hence a count for them is required.
+	// After enough writes are blocked, we want to notify VTOrc to run an ERS.
+	inProgressWriteCount int
+	// isBlocked stores if the primary is blocked on semi-sync ack.
+	isBlocked bool
+	// waiters stores the list of waiters that are waiting for the primary to be unblocked.
+	waiters []chan any
+	// writesBlockedGauge is a gauge tracking the number of writes the monitor is blocked on.
+	writesBlockedGauge *stats.Gauge
+}
+
+// NewMonitor creates a new Monitor.
+func NewMonitor(config *tabletenv.TabletConfig, exporter *servenv.Exporter) *Monitor {
+	return &Monitor{
+		config: config,
+		ticks:  timer.NewTimer(config.SemiSyncMonitor.Interval),
+		// We clear the data every day. We can make it configurable in the future,
+		// but this seams fine for now.
+		clearTicks:         timer.NewTimer(clearTimerDuration),
+		writesBlockedGauge: exporter.NewGauge("SemiSyncMonitorWritesBlocked", "Number of writes blocked in the semi-sync monitor"),
+		appPool:            dbconnpool.NewConnectionPool("SemiSyncMonitorAppPool", exporter, 20, mysqlctl.DbaIdleTimeout, 0, mysqlctl.PoolDynamicHostnameResolution),
+		waiters:            make([]chan any, 0),
+	}
+}
+
+// CreateTestSemiSyncMonitor created a monitor for testing.
+// It takes an optional fake db.
+func CreateTestSemiSyncMonitor(db *fakesqldb.DB, exporter *servenv.Exporter) *Monitor {
+	var dbc *dbconfigs.DBConfigs
+	if db != nil {
+		params := db.ConnParams()
+		cp := *params
+		dbc = dbconfigs.NewTestDBConfigs(cp, cp, "")
+	}
+	return NewMonitor(&tabletenv.TabletConfig{
+		DB: dbc,
+		SemiSyncMonitor: tabletenv.SemiSyncMonitorConfig{
+			Interval: 1 * time.Second,
+		},
+	}, exporter)
+}
+
+// Open starts the monitor.
+func (m *Monitor) Open() {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	// The check for config being nil is only requried for tests.
+	if m.isOpen || m.config == nil || m.config.DB == nil {
+		// If we are already open, then there is nothing to do
+		return
+	}
+	// Set the monitor to be open.
+	m.isOpen = true
+	log.Info("SemiSync Monitor: opening")
+
+	// This function could be running from within a unit test scope, in which case we use
+	// mock pools that are already open. This is why we test for the pool being open.
+	if !m.appPool.IsOpen() {
+		m.appPool.Open(m.config.DB.AppWithDB())
+	}
+	m.clearTicks.Start(m.clearAllData)
```

Comment:
What happens if we are in the middle of fixing a blocked semi-sync and this timer goes off?

Reply:
The write to clear the table would block if there is an outstanding write to the table in progress, and eventually it would succeed. There should be no problems with this triggering in the middle of fixing of semi-sync.